File: clients/ios/Lextures/Core/LMS/ContentToolPack4Logic.swift
// swiftlint:disable identifier_name large_tuple force_try type_body_length file_length
import Foundation

/// Pure CT.M8 pack-4 decisions — checkpoint scheduling/seek clamp, resume position,
/// worked-example step machine, slider throttle/settle, and client allowlist. No networking.
enum ContentToolPack4Logic {
    /// Native pack-4 tools (registered in CT.M3 registry).
    static let pack4ToolIds: Set<String> = [
        "media_checkpoints",
        "worked_example",
        "parameter_explorer",
    ]

    /// Delivered via CT.M4 sandbox only — never native-registered.
    static let sandboxToolIds: Set<String> = [
        "code_sandbox",
    ]

    /// Per-tool client allowlist (rollout). Empty entry removes a renderer without a release.
    static var clientAllowlist: Set<String> = pack4ToolIds

    static let checkpointToleranceSec = 0.25
    static let progressThrottleMs: Int64 = 5_000
    static let recomputeThrottleMs: Int64 = 33 // ≤ 30 Hz
    static let defaultCheckpointAttempts = 2
    static let defaultAttemptsPerStep = 3
    static let segmentGranularitySec = 5.0

    // MARK: - Types

    struct Checkpoint: Equatable {
        var id: String
        var atSec: Double
        var required: Bool
        var attempts: Int
    }

    struct CheckpointAnswer: Equatable {
        var done: Bool
        var attemptCount: Int
        var lastCorrect: Bool
    }

    struct SeekClamp: Equatable {
        var time: Double
        var clamped: Bool
    }

    enum StepStatus: String, Equatable {
        case current
        case solved
        case revealed
        case locked
        case scaffolded
        case allComplete = "all_complete"
    }

    struct AnswerResultView: Equatable {
        var correct: Bool?
        var feedback: String?
        var attemptsRemaining: Int?
        var done: Bool?
        var error: String?
        var message: String?
        var checkpointId: String?
    }

    struct CheckStepResultView: Equatable {
        var result: String?
        var feedback: String?
        var attemptsRemaining: Int?
        var canReveal: Bool?
        var nextStep: String?
        var stepId: String?
        var error: String?
        var message: String?
    }

    struct HintResultView: Equatable {
        var hint: String?
        var hintsRemaining: Int?
        var level: Int?
        var noMoreHints: Bool?
        var stepId: String?
        var error: String?
    }

    // MARK: - Allowlist / registry

    static func isClientAllowlisted(
        _ toolId: String,
        allowlist: Set<String> = clientAllowlist
    ) -> Bool {
        allowlist.contains(toolId)
    }

    static func allowlistedToolIds(allowlist: Set<String> = clientAllowlist) -> Set<String> {
        pack4ToolIds.intersection(allowlist)
    }

    static func conflictPolicy(for toolId: String) -> ContentToolHostLogic.ConflictPolicy {
        // All pack-4 tools (native + sandbox) use server_wins.
        switch toolId {
        case "media_checkpoints", "worked_example", "parameter_explorer", "code_sandbox":
            return .serverWins
        default:
            return .serverWins
        }
    }

    /// Pack-4 actions are never queued offline (CT.M3 FR-11).
    static func canQueueActionOffline(toolId: String, action: String) -> Bool {
        _ = toolId
        _ = action
        return false
    }

    // MARK: - Media checkpoints

    static func parseCheckpoints(_ config: JSONValue?) -> [Checkpoint] {
        arrayField(config, key: "checkpoints").compactMap { raw in
            let obj = objectMap(raw)
            guard case .string(let id) = obj["id"] else { return nil }
            let atSec = numberField(raw, key: "atSec") ?? 0
            let required = boolField(raw, key: "required") ?? true
            let attemptsRaw = numberField(raw, key: "attempts").map { Int($0.rounded()) }
                ?? defaultCheckpointAttempts
            let attempts = min(10, max(1, attemptsRaw))
            return Checkpoint(id: id, atSec: atSec, required: required, attempts: attempts)
        }
        .sorted { $0.atSec < $1.atSec }
    }

    static func parseAnswers(_ state: JSONValue?) -> [String: CheckpointAnswer] {
        let answersObj = objectMap(objectMap(state)["answers"])
        var out: [String: CheckpointAnswer] = [:]
        for (id, raw) in answersObj {
            let obj = objectMap(raw)
            let done = boolField(raw, key: "done") ?? false
            let attempts = arrayField(raw, key: "attempts")
            let lastCorrect: Bool
            if let last = attempts.last {
                lastCorrect = boolField(last, key: "correct") ?? false
            } else {
                lastCorrect = false
            }
            out[id] = CheckpointAnswer(
                done: done,
                attemptCount: attempts.count,
                lastCorrect: lastCorrect
            )
        }
        return out
    }

    static func isCheckpointDone(answers: [String: CheckpointAnswer], checkpoint: Checkpoint) -> Bool {
        guard let ans = answers[checkpoint.id] else { return false }
        if ans.done { return true }
        if ans.lastCorrect { return true }
        return ans.attemptCount >= checkpoint.attempts
    }

    static func findDueCheckpoint(
        checkpoints: [Checkpoint],
        answers: [String: CheckpointAnswer],
        currentTime: Double,
        alreadyPromptedIds: Set<String>
    ) -> Checkpoint? {
        for cp in checkpoints {
            if alreadyPromptedIds.contains(cp.id) { continue }
            if isCheckpointDone(answers: answers, checkpoint: cp) { continue }
            if currentTime + checkpointToleranceSec >= cp.atSec && currentTime + 2 >= cp.atSec {
                if currentTime >= cp.atSec - checkpointToleranceSec {
                    return cp
                }
            }
        }
        return nil
    }

    static func earliestUnansweredRequiredSec(
        checkpoints: [Checkpoint],
        answers: [String: CheckpointAnswer]
    ) -> Double? {
        var earliest: Double?
        for cp in checkpoints {
            if !cp.required { continue }
            if isCheckpointDone(answers: answers, checkpoint: cp) { continue }
            if earliest == nil || cp.atSec < earliest! {
                earliest = cp.atSec
            }
        }
        return earliest
    }

    static func clampSeekTime(
        preventSkip: Bool,
        checkpoints: [Checkpoint],
        answers: [String: CheckpointAnswer],
        targetSec: Double
    ) -> SeekClamp {
        guard preventSkip else { return SeekClamp(time: targetSec, clamped: false) }
        guard let limit = earliestUnansweredRequiredSec(checkpoints: checkpoints, answers: answers) else {
            return SeekClamp(time: targetSec, clamped: false)
        }
        if targetSec > limit + 0.05 {
            return SeekClamp(time: limit, clamped: true)
        }
        return SeekClamp(time: targetSec, clamped: false)
    }

    static func mergeLocalSegments(
        existing: [[Double]],
        start: Double,
        end: Double,
        granularity: Double = segmentGranularitySec
    ) -> [[Double]] {
        guard end > start else { return existing }
        func floorG(_ v: Double) -> Double {
            floor(max(0, v) / granularity) * granularity
        }
        func ceilG(_ v: Double) -> Double {
            let f = floorG(v)
            return v > f ? f + granularity : f
        }
        var s = floorG(start)
        var e = ceilG(end)
        if e <= s { e = s + granularity }
        var all = existing + [[s, e]]
        all.sort { a, b in
            if a[0] != b[0] { return a[0] < b[0] }
            return a[1] < b[1]
        }
        var merged: [[Double]] = []
        for seg in all {
            if let last = merged.last, seg[0] <= last[1] {
                if seg[1] > last[1] {
                    merged[merged.count - 1] = [last[0], seg[1]]
                }
            } else {
                merged.append(seg)
            }
        }
        return merged
    }

    static func resumePosition(furthestSec: Double?, watchedSegments: [[Double]]) -> Double {
        var best = furthestSec ?? 0
        for seg in watchedSegments {
            if seg.count >= 2 {
                best = max(best, seg[1])
            }
        }
        return max(0, best)
    }

    static func shouldFireProgressThrottle(lastFiredAtMs: Int64?, nowMs: Int64, intervalMs: Int64 = progressThrottleMs) -> Bool {
        guard let last = lastFiredAtMs else { return true }
        return nowMs - last >= intervalMs
    }

    static func shouldBlockPlayback(
        checkpoint: Checkpoint,
        answers: [String: CheckpointAnswer]
    ) -> Bool {
        guard checkpoint.required else { return false }
        return !isCheckpointDone(answers: answers, checkpoint: checkpoint)
    }

    /// Direct media / course files are reliable; YouTube/Vimeo embeds are not for checkpoints.
    static func hasReliableCheckpointTiming(
        source: String?,
        url: String?,
        provider: String?
    ) -> Bool {
        let trimmed = url?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        guard !trimmed.isEmpty else { return false }
        let p = (provider ?? "").lowercased()
        if p == "youtube" || p == "vimeo" { return false }
        let lower = trimmed.lowercased()
        if lower.contains("youtube.com") || lower.contains("youtu.be") || lower.contains("vimeo.com") {
            return false
        }
        _ = source
        return true
    }

    static func parseWatchedSegments(_ state: JSONValue?) -> [[Double]] {
        arrayField(state, key: "watchedSegments").compactMap { raw in
            guard case .array(let arr) = raw, arr.count >= 2 else { return nil }
            guard case .number(let a) = arr[0], case .number(let b) = arr[1] else { return nil }
            return [a, b]
        }
    }

    static func parseAnswerResult(_ value: JSONValue?) -> AnswerResultView {
        AnswerResultView(
            correct: boolField(value, key: "correct"),
            feedback: stringField(value, key: "feedback"),
            attemptsRemaining: numberField(value, key: "attemptsRemaining").map { Int($0.rounded()) },
            done: boolField(value, key: "done"),
            error: stringField(value, key: "error"),
            message: stringField(value, key: "message"),
            checkpointId: stringField(value, key: "checkpointId")
        )
    }

    // MARK: - Worked example

    static func parseStepIds(_ config: JSONValue?) -> [String] {
        arrayField(config, key: "steps").compactMap { raw in
            stringField(raw, key: "id")
        }
    }

    static func blankedStepIds(config: JSONValue?, state: JSONValue?) -> [String] {
        let fromState = arrayField(state, key: "blankedStepIds").compactMap { raw -> String? in
            if case .string(let s) = raw { return s }
            return nil
        }
        if !fromState.isEmpty { return fromState }
        return arrayField(config, key: "steps").compactMap { raw in
            let obj = objectMap(raw)
            guard obj["blank"] != nil, case .string(let id) = obj["id"] else { return nil }
            return id
        }
    }

    static func isStepDone(_ progress: JSONValue?) -> Bool {
        if boolField(progress, key: "revealed") == true { return true }
        if stringField(progress, key: "completedAt") != nil { return true }
        let attempts = arrayField(progress, key: "attempts")
        guard let last = attempts.last else { return false }
        let result = stringField(last, key: "result")
        return result == "correct" || result == "needs_review"
    }

    static func stepProgressMap(_ state: JSONValue?) -> [String: JSONValue] {
        objectMap(objectMap(state)["steps"])
    }

    static func resolveCurrentStepId(
        blankedStepIds: [String],
        currentStepId: String?,
        progress: [String: JSONValue]
    ) -> String {
        if let currentStepId, !currentStepId.isEmpty {
            return currentStepId
        }
        for id in blankedStepIds {
            if !isStepDone(progress[id]) { return id }
        }
        return blankedStepIds.first ?? ""
    }

    static func stepStatus(
        stepId: String,
        blankedStepIds: [String],
        currentStepId: String,
        progress: [String: JSONValue],
        allStepIds: [String]
    ) -> StepStatus {
        let blanked = Set(blankedStepIds)
        if !blanked.contains(stepId) {
            return .scaffolded
        }
        let sp = progress[stepId]
        if boolField(sp, key: "revealed") == true {
            return .revealed
        }
        if isStepDone(sp) {
            return .solved
        }
        if stepId == currentStepId {
            return .current
        }
        let currentIndex = blankedStepIds.firstIndex(of: currentStepId) ?? 0
        let stepIndex = blankedStepIds.firstIndex(of: stepId) ?? Int.max
        if stepIndex > currentIndex {
            return .locked
        }
        _ = allStepIds
        return .locked
    }

    static func canCheckStep(draft: String, readOnly: Bool, busy: Bool, stepDone: Bool) -> Bool {
        if readOnly || busy || stepDone { return false }
        return !draft.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
    }

    static func mergeStepDraft(
        state: [String: JSONValue],
        stepId: String,
        draft: String
    ) -> [String: JSONValue] {
        var steps = objectMap(state["steps"])
        var sp = objectMap(steps[stepId])
        sp["draft"] = .string(draft)
        steps[stepId] = .object(sp)
        return mergePreservingUnknown(base: state, patch: [
            "v": state["v"] ?? .number(1),
            "steps": .object(steps),
        ])
    }

    static func parseCheckStepResult(_ value: JSONValue?) -> CheckStepResultView {
        CheckStepResultView(
            result: stringField(value, key: "result"),
            feedback: stringField(value, key: "feedback"),
            attemptsRemaining: numberField(value, key: "attemptsRemaining").map { Int($0.rounded()) },
            canReveal: boolField(value, key: "canReveal"),
            nextStep: stringField(value, key: "nextStep"),
            stepId: stringField(value, key: "stepId"),
            error: stringField(value, key: "error"),
            message: stringField(value, key: "message")
        )
    }

    static func parseHintResult(_ value: JSONValue?) -> HintResultView {
        HintResultView(
            hint: stringField(value, key: "hint"),
            hintsRemaining: numberField(value, key: "hintsRemaining").map { Int($0.rounded()) },
            level: numberField(value, key: "level").map { Int($0.rounded()) },
            noMoreHints: boolField(value, key: "noMoreHints"),
            stepId: stringField(value, key: "stepId"),
            error: stringField(value, key: "error")
        )
    }

    // MARK: - Parameter explorer

    static func shouldRecompute(lastAtMs: Int64?, nowMs: Int64, intervalMs: Int64 = recomputeThrottleMs) -> Bool {
        guard let last = lastAtMs else { return true }
        return nowMs - last >= intervalMs
    }

    static func shouldAutosaveOnSettle(dragging: Bool, dirty: Bool) -> Bool {
        !dragging && dirty
    }

    static func defaultParams(from config: JSONValue?) -> [String: JSONValue] {
        var out: [String: JSONValue] = [:]
        for raw in arrayField(config, key: "parameters") {
            let obj = objectMap(raw)
            guard case .string(let id) = obj["id"], let def = obj["default"] else { continue }
            out[id] = def
        }
        return out
    }

    static func clampNumber(value: Double, min: Double, max: Double, step: Double) -> Double {
        let lo = Swift.min(min, max)
        let hi = Swift.max(min, max)
        var v = Swift.max(lo, Swift.min(hi, value))
        if step > 0 {
            let steps = ((v - lo) / step).rounded()
            v = lo + steps * step
            v = Swift.max(lo, Swift.min(hi, v))
        }
        return v
    }

    static func mergeParamsPreservingUnknown(
        state: [String: JSONValue],
        params: [String: JSONValue]
    ) -> [String: JSONValue] {
        mergePreservingUnknown(base: state, patch: [
            "v": state["v"] ?? .number(1),
            "params": .object(params),
        ])
    }

    // MARK: - JSON helpers

    static func mergePreservingUnknown(
        base: [String: JSONValue],
        patch: [String: JSONValue]
    ) -> [String: JSONValue] {
        var out = base
        for (k, v) in patch { out[k] = v }
        return out
    }

    static func objectMap(_ value: JSONValue?) -> [String: JSONValue] {
        guard case .object(let obj) = value else { return [:] }
        return obj
    }

    static func arrayField(_ value: JSONValue?, key: String) -> [JSONValue] {
        guard case .object(let obj) = value, case .array(let arr) = obj[key] else { return [] }
        return arr
    }

    static func boolField(_ value: JSONValue?, key: String) -> Bool? {
        guard case .object(let obj) = value, let field = obj[key] else { return nil }
        if case .bool(let flag) = field { return flag }
        return nil
    }

    static func numberField(_ value: JSONValue?, key: String) -> Double? {
        guard case .object(let obj) = value, let field = obj[key] else { return nil }
        if case .number(let number) = field { return number }
        if case .string(let text) = field { return Double(text) }
        return nil
    }

    static func stringField(_ value: JSONValue?, key: String) -> String? {
        ContentToolHostLogic.stringField(value, key: key)
    }
}

File: clients/android/app/src/main/kotlin/com/lextures/android/core/lms/ContentToolPack4Logic.kt
package com.lextures.android.core.lms

import kotlin.math.floor
import kotlin.math.max
import kotlin.math.min
import kotlin.math.round
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.booleanOrNull
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.doubleOrNull
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive

/**
 * Pure CT.M8 pack-4 decisions — checkpoint scheduling/seek clamp, resume position,
 * worked-example step machine, slider throttle/settle, and client allowlist. No networking.
 */
object ContentToolPack4Logic {
    /** Native pack-4 tools (registered in CT.M3 registry). */
    val pack4ToolIds: Set<String> = setOf(
        "media_checkpoints",
        "worked_example",
        "parameter_explorer",
    )

    /** Delivered via CT.M4 sandbox only — never native-registered. */
    val sandboxToolIds: Set<String> = setOf(
        "code_sandbox",
    )

    /** Per-tool client allowlist (rollout). Empty entry removes a renderer without a release. */
    var clientAllowlist: Set<String> = pack4ToolIds

    const val CHECKPOINT_TOLERANCE_SEC = 0.25
    const val PROGRESS_THROTTLE_MS: Long = 5_000
    const val RECOMPUTE_THROTTLE_MS: Long = 33 // ≤ 30 Hz
    const val DEFAULT_CHECKPOINT_ATTEMPTS = 2
    const val DEFAULT_ATTEMPTS_PER_STEP = 3
    const val SEGMENT_GRANULARITY_SEC = 5.0

    data class Checkpoint(
        val id: String,
        val atSec: Double,
        val required: Boolean,
        val attempts: Int,
    )

    data class CheckpointAnswer(
        val done: Boolean,
        val attemptCount: Int,
        val lastCorrect: Boolean,
    )

    data class SeekClamp(
        val time: Double,
        val clamped: Boolean,
    )

    enum class StepStatus(val wire: String) {
        CURRENT("current"),
        SOLVED("solved"),
        REVEALED("revealed"),
        LOCKED("locked"),
        SCAFFOLDED("scaffolded"),
        ALL_COMPLETE("all_complete"),
    }

    data class AnswerResultView(
        val correct: Boolean?,
        val feedback: String?,
        val attemptsRemaining: Int?,
        val done: Boolean?,
        val error: String?,
        val message: String?,
        val checkpointId: String?,
    )

    data class CheckStepResultView(
        val result: String?,
        val feedback: String?,
        val attemptsRemaining: Int?,
        val canReveal: Boolean?,
        val nextStep: String?,
        val stepId: String?,
        val error: String?,
        val message: String?,
    )

    data class HintResultView(
        val hint: String?,
        val hintsRemaining: Int?,
        val level: Int?,
        val noMoreHints: Boolean?,
        val stepId: String?,
        val error: String?,
    )

    // MARK: - Allowlist / registry

    fun isClientAllowlisted(
        toolId: String,
        allowlist: Set<String> = clientAllowlist,
    ): Boolean = toolId in allowlist

    fun allowlistedToolIds(allowlist: Set<String> = clientAllowlist): Set<String> =
        pack4ToolIds.intersect(allowlist)

    fun conflictPolicy(toolId: String): ContentToolHostLogic.ConflictPolicy =
        when (toolId) {
            "media_checkpoints", "worked_example", "parameter_explorer", "code_sandbox" ->
                ContentToolHostLogic.ConflictPolicy.SERVER_WINS
            else -> ContentToolHostLogic.ConflictPolicy.SERVER_WINS
        }

    /** Pack-4 actions are never queued offline (CT.M3 FR-11). */
    @Suppress("UNUSED_PARAMETER")
    fun canQueueActionOffline(toolId: String, action: String): Boolean = false

    // MARK: - Media checkpoints

    fun parseCheckpoints(config: JsonElement?): List<Checkpoint> =
        arrayField(config, "checkpoints").mapNotNull { raw ->
            val obj = objectMap(raw)
            val id = (obj["id"] as? JsonPrimitive)?.contentOrNull ?: return@mapNotNull null
            val atSec = numberField(raw, "atSec") ?: 0.0
            val required = boolField(raw, "required") ?: true
            val attemptsRaw = numberField(raw, "attempts")?.toInt() ?: DEFAULT_CHECKPOINT_ATTEMPTS
            val attempts = min(10, max(1, attemptsRaw))
            Checkpoint(id, atSec, required, attempts)
        }.sortedBy { it.atSec }

    fun parseAnswers(state: JsonElement?): Map<String, CheckpointAnswer> {
        val answersObj = objectMap(objectMap(state)["answers"])
        return answersObj.mapValues { (_, raw) ->
            val attempts = arrayField(raw, "attempts")
            val lastCorrect = attempts.lastOrNull()?.let { boolField(it, "correct") } ?: false
            CheckpointAnswer(
                done = boolField(raw, "done") ?: false,
                attemptCount = attempts.size,
                lastCorrect = lastCorrect,
            )
        }
    }

    fun isCheckpointDone(answers: Map<String, CheckpointAnswer>, checkpoint: Checkpoint): Boolean {
        val ans = answers[checkpoint.id] ?: return false
        if (ans.done) return true
        if (ans.lastCorrect) return true
        return ans.attemptCount >= checkpoint.attempts
    }

    fun findDueCheckpoint(
        checkpoints: List<Checkpoint>,
        answers: Map<String, CheckpointAnswer>,
        currentTime: Double,
        alreadyPromptedIds: Set<String>,
    ): Checkpoint? {
        for (cp in checkpoints) {
            if (cp.id in alreadyPromptedIds) continue
            if (isCheckpointDone(answers, cp)) continue
            if (currentTime + CHECKPOINT_TOLERANCE_SEC >= cp.atSec && currentTime + 2 >= cp.atSec) {
                if (currentTime >= cp.atSec - CHECKPOINT_TOLERANCE_SEC) {
                    return cp
                }
            }
        }
        return null
    }

    fun earliestUnansweredRequiredSec(
        checkpoints: List<Checkpoint>,
        answers: Map<String, CheckpointAnswer>,
    ): Double? {
        var earliest: Double? = null
        for (cp in checkpoints) {
            if (!cp.required) continue
            if (isCheckpointDone(answers, cp)) continue
            if (earliest == null || cp.atSec < earliest) earliest = cp.atSec
        }
        return earliest
    }

    fun clampSeekTime(
        preventSkip: Boolean,
        checkpoints: List<Checkpoint>,
        answers: Map<String, CheckpointAnswer>,
        targetSec: Double,
    ): SeekClamp {
        if (!preventSkip) return SeekClamp(targetSec, false)
        val limit = earliestUnansweredRequiredSec(checkpoints, answers) ?: return SeekClamp(targetSec, false)
        if (targetSec > limit + 0.05) return SeekClamp(limit, true)
        return SeekClamp(targetSec, false)
    }

    fun mergeLocalSegments(
        existing: List<List<Double>>,
        start: Double,
        end: Double,
        granularity: Double = SEGMENT_GRANULARITY_SEC,
    ): List<List<Double>> {
        if (end <= start) return existing
        fun floorG(v: Double): Double = floor(max(0.0, v) / granularity) * granularity
        fun ceilG(v: Double): Double {
            val f = floorG(v)
            return if (v > f) f + granularity else f
        }
        var s = floorG(start)
        var e = ceilG(end)
        if (e <= s) e = s + granularity
        val all = (existing + listOf(listOf(s, e))).sortedWith(compareBy({ it[0] }, { it[1] }))
        val merged = mutableListOf<List<Double>>()
        for (seg in all) {
            val last = merged.lastOrNull()
            if (last == null || seg[0] > last[1]) {
                merged += seg
            } else if (seg[1] > last[1]) {
                merged[merged.lastIndex] = listOf(last[0], seg[1])
            }
        }
        return merged
    }

    fun resumePosition(furthestSec: Double?, watchedSegments: List<List<Double>>): Double {
        var best = furthestSec ?: 0.0
        for (seg in watchedSegments) {
            if (seg.size >= 2) best = max(best, seg[1])
        }
        return max(0.0, best)
    }

    fun shouldFireProgressThrottle(
        lastFiredAtMs: Long?,
        nowMs: Long,
        intervalMs: Long = PROGRESS_THROTTLE_MS,
    ): Boolean {
        if (lastFiredAtMs == null) return true
        return nowMs - lastFiredAtMs >= intervalMs
    }

    fun shouldBlockPlayback(
        checkpoint: Checkpoint,
        answers: Map<String, CheckpointAnswer>,
    ): Boolean {
        if (!checkpoint.required) return false
        return !isCheckpointDone(answers, checkpoint)
    }

    /** Direct media / course files are reliable; YouTube/Vimeo embeds are not for checkpoints. */
    fun hasReliableCheckpointTiming(
        source: String?,
        url: String?,
        provider: String?,
    ): Boolean {
        val trimmed = url?.trim().orEmpty()
        if (trimmed.isEmpty()) return false
        val p = provider.orEmpty().lowercase()
        if (p == "youtube" || p == "vimeo") return false
        val lower = trimmed.lowercase()
        if ("youtube.com" in lower || "youtu.be" in lower || "vimeo.com" in lower) return false
        @Suppress("UNUSED_EXPRESSION")
        source
        return true
    }

    fun parseWatchedSegments(state: JsonElement?): List<List<Double>> =
        arrayField(state, "watchedSegments").mapNotNull { raw ->
            val arr = raw as? JsonArray ?: return@mapNotNull null
            if (arr.size < 2) return@mapNotNull null
            val a = (arr[0] as? JsonPrimitive)?.doubleOrNull ?: return@mapNotNull null
            val b = (arr[1] as? JsonPrimitive)?.doubleOrNull ?: return@mapNotNull null
            listOf(a, b)
        }

    fun parseAnswerResult(value: JsonElement?): AnswerResultView =
        AnswerResultView(
            correct = boolField(value, "correct"),
            feedback = stringField(value, "feedback"),
            attemptsRemaining = numberField(value, "attemptsRemaining")?.toInt(),
            done = boolField(value, "done"),
            error = stringField(value, "error"),
            message = stringField(value, "message"),
            checkpointId = stringField(value, "checkpointId"),
        )

    // MARK: - Worked example

    fun parseStepIds(config: JsonElement?): List<String> =
        arrayField(config, "steps").mapNotNull { stringField(it, "id") }

    fun blankedStepIds(config: JsonElement?, state: JsonElement?): List<String> {
        val fromState = arrayField(state, "blankedStepIds").mapNotNull {
            (it as? JsonPrimitive)?.contentOrNull
        }
        if (fromState.isNotEmpty()) return fromState
        return arrayField(config, "steps").mapNotNull { raw ->
            val obj = objectMap(raw)
            if (obj["blank"] == null) return@mapNotNull null
            (obj["id"] as? JsonPrimitive)?.contentOrNull
        }
    }

    fun isStepDone(progress: JsonElement?): Boolean {
        if (boolField(progress, "revealed") == true) return true
        if (stringField(progress, "completedAt") != null) return true
        val attempts = arrayField(progress, "attempts")
        val last = attempts.lastOrNull() ?: return false
        val result = stringField(last, "result")
        return result == "correct" || result == "needs_review"
    }

    fun stepProgressMap(state: JsonElement?): Map<String, JsonElement> =
        objectMap(objectMap(state)["steps"])

    fun resolveCurrentStepId(
        blankedStepIds: List<String>,
        currentStepId: String?,
        progress: Map<String, JsonElement>,
    ): String {
        if (!currentStepId.isNullOrEmpty()) return currentStepId
        for (id in blankedStepIds) {
            if (!isStepDone(progress[id])) return id
        }
        return blankedStepIds.firstOrNull().orEmpty()
    }

    fun stepStatus(
        stepId: String,
        blankedStepIds: List<String>,
        currentStepId: String,
        progress: Map<String, JsonElement>,
        allStepIds: List<String>,
    ): StepStatus {
        val blanked = blankedStepIds.toSet()
        if (stepId !in blanked) return StepStatus.SCAFFOLDED
        val sp = progress[stepId]
        if (boolField(sp, "revealed") == true) return StepStatus.REVEALED
        if (isStepDone(sp)) return StepStatus.SOLVED
        if (stepId == currentStepId) return StepStatus.CURRENT
        @Suppress("UNUSED_EXPRESSION")
        allStepIds
        return StepStatus.LOCKED
    }

    fun canCheckStep(draft: String, readOnly: Boolean, busy: Boolean, stepDone: Boolean): Boolean {
        if (readOnly || busy || stepDone) return false
        return draft.trim().isNotEmpty()
    }

    fun mergeStepDraft(
        state: Map<String, JsonElement>,
        stepId: String,
        draft: String,
    ): Map<String, JsonElement> {
        val steps = objectMap(state["steps"]).toMutableMap()
        val sp = objectMap(steps[stepId]).toMutableMap()
        sp["draft"] = JsonPrimitive(draft)
        steps[stepId] = JsonObject(sp)
        return mergePreservingUnknown(
            state,
            mapOf(
                "v" to (state["v"] ?: JsonPrimitive(1)),
                "steps" to JsonObject(steps),
            ),
        )
    }

    fun parseCheckStepResult(value: JsonElement?): CheckStepResultView =
        CheckStepResultView(
            result = stringField(value, "result"),
            feedback = stringField(value, "feedback"),
            attemptsRemaining = numberField(value, "attemptsRemaining")?.toInt(),
            canReveal = boolField(value, "canReveal"),
            nextStep = stringField(value, "nextStep"),
            stepId = stringField(value, "stepId"),
            error = stringField(value, "error"),
            message = stringField(value, "message"),
        )

    fun parseHintResult(value: JsonElement?): HintResultView =
        HintResultView(
            hint = stringField(value, "hint"),
            hintsRemaining = numberField(value, "hintsRemaining")?.toInt(),
            level = numberField(value, "level")?.toInt(),
            noMoreHints = boolField(value, "noMoreHints"),
            stepId = stringField(value, "stepId"),
            error = stringField(value, "error"),
        )

    // MARK: - Parameter explorer

    fun shouldRecompute(
        lastAtMs: Long?,
        nowMs: Long,
        intervalMs: Long = RECOMPUTE_THROTTLE_MS,
    ): Boolean {
        if (lastAtMs == null) return true
        return nowMs - lastAtMs >= intervalMs
    }

    fun shouldAutosaveOnSettle(dragging: Boolean, dirty: Boolean): Boolean =
        !dragging && dirty

    fun defaultParams(config: JsonElement?): Map<String, JsonElement> {
        val out = mutableMapOf<String, JsonElement>()
        for (raw in arrayField(config, "parameters")) {
            val obj = objectMap(raw)
            val id = (obj["id"] as? JsonPrimitive)?.contentOrNull ?: continue
            val def = obj["default"] ?: continue
            out[id] = def
        }
        return out
    }

    fun clampNumber(value: Double, minV: Double, maxV: Double, step: Double): Double {
        val lo = min(minV, maxV)
        val hi = max(minV, maxV)
        var v = max(lo, min(hi, value))
        if (step > 0) {
            val steps = round((v - lo) / step)
            v = lo + steps * step
            v = max(lo, min(hi, v))
        }
        return v
    }

    fun mergeParamsPreservingUnknown(
        state: Map<String, JsonElement>,
        params: Map<String, JsonElement>,
    ): Map<String, JsonElement> =
        mergePreservingUnknown(
            state,
            mapOf(
                "v" to (state["v"] ?: JsonPrimitive(1)),
                "params" to JsonObject(params),
            ),
        )

    // MARK: - JSON helpers

    fun mergePreservingUnknown(
        base: Map<String, JsonElement>,
        patch: Map<String, JsonElement>,
    ): Map<String, JsonElement> = base + patch

    fun objectMap(value: JsonElement?): Map<String, JsonElement> =
        value?.jsonObject ?: emptyMap()

    fun arrayField(value: JsonElement?, key: String): List<JsonElement> {
        val obj = value?.jsonObject ?: return emptyList()
        return obj[key]?.jsonArray?.toList().orEmpty()
    }

    fun boolField(value: JsonElement?, key: String): Boolean? {
        val obj = value?.jsonObject ?: return null
        return (obj[key] as? JsonPrimitive)?.booleanOrNull
    }

    fun numberField(value: JsonElement?, key: String): Double? {
        val obj = value?.jsonObject ?: return null
        val field = obj[key] ?: return null
        val prim = field as? JsonPrimitive ?: return null
        prim.doubleOrNull?.let { return it }
        return prim.contentOrNull?.toDoubleOrNull()
    }

    fun stringField(value: JsonElement?, key: String): String? =
        ContentToolHostLogic.stringField(value, key)
}
